### PR TITLE
Roundtrip Optimization For Binary Encoding

### PR DESCRIPTION
**Pull request by bchong95**
# Roundtrip Optimization For Binary Encoding

## Description

Enabling same format round trip for binary encoding.

## Type of change

Please delete options that are not relevant.

- [x] Bug fix (non-breaking change which fixes an issue)

### CODE REVIEW

File: Microsoft.Azure.Cosmos/src/Json/JsonWriter.JsonBinaryWriter.cs [hunk 1]
```diff
@@ -284,9 +285,37 @@ public override byte[] GetResult()
...
+                    case JsonTokenType.FieldName:
+                        break;
+                    default:
+                        throw new ArgumentException($"{nameof(JsonBinaryWriter)}.{nameof(WriteRawJsonToken)} can not write a {nameof(JsonTokenType)}: {jsonTokenType}");
+                }
+
+                this.JsonObjectState.RegisterToken(jsonTokenType);
+
+                foreach (byte byteValue in rawJsonToken)
+                {
```
Comment (jasontho-ms) [inline Microsoft.Azure.Cosmos/src/Json/JsonWriter.JsonBinaryWriter.cs]: Byte-by-byte will be too slow

File: Microsoft.Azure.Cosmos/src/Json/JsonWriter.JsonBinaryWriter.cs [hunk 2]
```diff
@@ -8,6 +8,7 @@ namespace Microsoft.Azure.Cosmos.Json
     using System;
```
Comment (jasontho-ms) [inline Microsoft.Azure.Cosmos/src/Json/JsonWriter.JsonBinaryWriter.cs]: Does JsonBinaryNavigator need to implement TryGetBufferedRawJson?

Comment (bchong95) [reply Microsoft.Azure.Cosmos/src/Json/JsonWriter.JsonBinaryWriter.cs]: Yeah somewhere else in the PR.

File: Microsoft.Azure.Cosmos/src/Json/JsonWriter.JsonBinaryWriter.cs (hunk 1 above)
Comment (bchong95) [reply Microsoft.Azure.Cosmos/src/Json/JsonWriter.JsonBinaryWriter.cs]: Agreed. I added the pattern matching check you suggested.

**Pull request merged**